VDA-Name-Registry/contracts/v2/NameRegistryV2.sol

```diff
@@ -0,0 +1,548 @@
+        }
+
+        AppMetaDataItem[] storage itemList = _didAppMetaData[did][_appName];
+        uint len = itemList.length;
+        bool isExistingItem;
+        for (uint i; i < len;) {
+            if (_isSameString(itemList[i].key, item.key)) {
+                isExistingItem = true;
+                itemList[i].value = item.value;
+                break;
+            }
+            unchecked {
+                ++i;
+            }
+        }
+        if (!isExistingItem) {
+            itemList.push(item);
+        }
+
+        emit UpdateApp(did, _ownerName, _appName, item);
+    }
+
+    /**
+     * @dev See {INameRegistryV2}
+     */
+    function getApp(string calldata ownerName, string calldata appName) external view virtual override returns(address, AppMetaDataItem[] memory) {
+        string memory _ownerName;
+        string memory _appName;
+        _ownerName = ownerName.lower();
+        _appName = appName.lower();
+
+
+        if (!_ownerNameInfo[_ownerName].registered) {
+            revert InvalidOwnerName();
+        }
+
+        address did;
+        did = _ownerNameInfo[_ownerName].DID;
+
+        if (!_didApps[did].contains(_appName)) {
+            revert InvalidAppName();
+        }
+
+        return (did, _didAppMetaData[did][_appName]);
+    }
+
+
+    /**
+     * @dev See {INameRegistryV2}
+     *      Once token address is set, it will never be set as zero address
+     */
+    function setTokenAddress(IERC20Upgradeable tokenAddr) external virtual payable override onlyOwner {
+        assembly {
+            if iszero(tokenAddr) {
+                let ptr := mload(0x40)
+                mstore(ptr, 0xaa7feadc00000000000000000000000000000000000000000000000000000000)
+                revert(ptr, 0x4) //revert InvalidValue()
+            }
+        }
+
+        if (address(vdaToken) == address(tokenAddr)) {
+            revert InvalidValue();
+        }
+
+        vdaToken = tokenAddr;
+
+        emit SetTokenAddress(address(tokenAddr));
+    }
+
+    /**
+     * @dev See {INameRegistryV2}
+     */
+    function getTokenAddress() external view virtual override returns(address) {
+        return address(vdaToken);
+    }
+
+    /**
+     * @dev See {INameRegistryV2}
+     *      Once fee is set, it will never be updated as 0
+     */
+    function updateAppRegisterFee(uint feeAmount) external virtual payable override onlyOwner {
+        if (feeAmount == appRegisterFee || feeAmount == 0) {
+            revert InvalidValue();
+        }
+
+        uint orgFee = appRegisterFee;
+        appRegisterFee = feeAmount;
+
+        emit UpdateAppRegisterFee(orgFee, appRegisterFee);
+    }
+
+    /**
+     * @dev See {INameRegistryV2}
+     */
+    function getAppRegisterFee() external view virtual override returns(uint) {
+        return appRegisterFee;
+    }
+
+    /**
+     * @dev See {INameRegistryV2}
+     */
+    function enableAppRegister(bool isEnabled) external virtual payable override onlyOwner {
```

Comment:
Need `disableAppRegister`

Reply:
As we've already deployed the `NameRegistry` contract on the testnet and mainnet, we should upgrade the current contracts. Here, the registering app feature is now disabled in the original `NameRegistry` contract.
After we upgrade the contract, we should set the fee for registering app and set the token address. 
And then, we should enable the app registering by call this `enableAppRegister(true)` function.

So, I think the function name of `enableAppRegister()` might be suitable.

Reply:
I agree with the name, I'm saying we should have an extra method to disable app registration. We may want to do that in the future.

Reply:
Once we called `enableAppRegister(false)`, it would disable the registering app.

Reply:
Oh I see, sorry I missed the boolean.

In that case, make it `isAppRegisterEnabled()`

Reply:
There is already `isAppRegisterEnabled()` function that returns the boolean.
![image](https://github.com/user-attachments/assets/b54d93c5-5fbd-4f6e-86ca-21b29159431f)

Reply:
Haha ok.

Let's call it `setAppRegisterEnabled()`

Reply:
Updated